Libraries/NativeAnimation/Drivers/RCTSpringAnimation.m

```diff
@@ -0,0 +1,209 @@
+    _callback = [callback copy];
+
+    _outputValue = _fromValue;
+    _lastPosition = _fromValue;
+    _lastVelocity = _initialVelocity;
+  }
+  return self;
+}
+
+RCT_NOT_IMPLEMENTED(- (instancetype)init)
+
+- (void)startAnimation
+{
+  _animationStartTime = CACurrentMediaTime();
+  _animationCurrentTime = _animationStartTime;
+  _animationHasBegun = YES;
+}
+
+- (void)stopAnimation
+{
+  _animationHasFinished = YES;
+}
+
+- (void)removeAnimation
+{
+  [self stopAnimation];
+  _valueNode = nil;
+  if (_callback) {
+    _callback(@[@{
+      @"finished": @(_animationHasFinished)
+    }]);
+  }
+}
+
+- (void)stepAnimation
+{
+  if (!_animationHasBegun || _animationHasFinished) {
+    // Animation has not begun or animation has already finished.
+    return;
+  }
+
+  // We are using a fixed time step and a maximum number of iterations.
+  // The following post provides a lot of thoughts into how to build this
+  // loop: http://gafferongames.com/game-physics/fix-your-timestep/
+  CGFloat TIMESTEP_MSEC = 1;
+  // Velocity is based on seconds instead of milliseconds
+  CGFloat step = TIMESTEP_MSEC / 1000;
+
+  NSTimeInterval currentTime = CACurrentMediaTime();
+  NSInteger numSteps = floorf((currentTime - _animationCurrentTime) / step);
+  _animationCurrentTime = currentTime;
+  if (numSteps == 0) {
+    return;
+  }
+
+  CGFloat position = _lastPosition;
+  CGFloat velocity = _lastVelocity;
+
+  CGFloat tempPosition = _lastPosition;
+  CGFloat tempVelocity = _lastVelocity;
+
+  for (NSInteger i = 0; i < numSteps; ++i) {
+    // This is using RK4. A good blog post to understand how it works:
+    // http://gafferongames.com/game-physics/integration-basics/
+    CGFloat aVelocity = velocity;
+    CGFloat aAcceleration = _tension * (_toValue - tempPosition) - _friction * tempVelocity;
+    tempPosition = position + aVelocity * step / 2;
+    tempVelocity = velocity + aAcceleration * step / 2;
+
+    CGFloat bVelocity = tempVelocity;
+    CGFloat bAcceleration = _tension * (_toValue - tempPosition) - _friction * tempVelocity;
+    tempPosition = position + bVelocity * step / 2;
+    tempVelocity = velocity + bAcceleration * step / 2;
+
+    CGFloat cVelocity = tempVelocity;
+    CGFloat cAcceleration = _tension * (_toValue - tempPosition) - _friction * tempVelocity;
+    tempPosition = position + cVelocity * step / 2;
+    tempVelocity = velocity + cAcceleration * step / 2;
+
+    CGFloat dVelocity = tempVelocity;
+    CGFloat dAcceleration = _tension * (_toValue - tempPosition) - _friction * tempVelocity;
+    tempPosition = position + cVelocity * step / 2;
+    tempVelocity = velocity + cAcceleration * step / 2;
+
+    CGFloat dxdt = (aVelocity + 2 * (bVelocity + cVelocity) + dVelocity) / 6;
+    CGFloat dvdt = (aAcceleration + 2 * (bAcceleration + cAcceleration) + dAcceleration) / 6;
+
+    position += dxdt * step;
+    velocity += dvdt * step;
+  }
+
+  _lastPosition = position;
+  _lastVelocity = velocity;
+
+  if (_animationHasFinished) {
```

Comment:
Is this check useful? As far as I can tell we don't update this value and we already check it at the start of the method, save for the one a few lines later and the one at the end.